File: typescript/src/environment.ts
import type { HttpClient } from "./http.js";
import type {
  EnvironmentData,
  ExecRequest,
  ExecResult,
  FileEntry,
  EnvironmentPort,
  CreatePortRequest,
} from "./types.js";

export class Environment implements EnvironmentData {
  readonly id: string;
  readonly org_id: string;
  readonly created_by: string;
  readonly template_id: string;
  readonly type: string;
  readonly name: string | undefined;
  readonly state: string;
  readonly desired_state: string;
  readonly error_reason: string | undefined;
  readonly recoverable: boolean;
  readonly state_changed_at: string;
  readonly retry_count: number;
  readonly ttl_seconds: number | undefined;
  readonly shutdown_at: string | undefined;
  readonly cpu_millicores: number;
  readonly memory_mb: number;
  readonly disk_gb: number;
  readonly total_running_seconds: number;
  readonly created_at: string;
  readonly started_at: string | undefined;
  readonly stopped_at: string | undefined;
  readonly destroyed_at: string | undefined;
  readonly ports: EnvironmentPort[] | undefined;
  readonly template_name: string | undefined;

  private http: HttpClient;
  private _orgId: string;

  constructor(data: EnvironmentData, http: HttpClient) {
    Object.assign(this, data);
    this.id = data.id;
    this.org_id = data.org_id;
    this.created_by = data.created_by;
    this.template_id = data.template_id;
    this.type = data.type;
    this.state = data.state;
    this.desired_state = data.desired_state;
    this.recoverable = data.recoverable;
    this.state_changed_at = data.state_changed_at;
    this.retry_count = data.retry_count;
    this.cpu_millicores = data.cpu_millicores;
    this.memory_mb = data.memory_mb;
    this.disk_gb = data.disk_gb;
    this.total_running_seconds = data.total_running_seconds;
    this.created_at = data.created_at;
    this.http = http;
    this._orgId = http.orgId;
  }

  private path(suffix: string): string {
    return `/organizations/${this._orgId}/environments/${this.id}${suffix}`;
  }

  async exec(req: ExecRequest): Promise<ExecResult> {
    return this.http.post(this.path("/exec"), req);
  }

  async start(): Promise<void> {
    await this.http.post(this.path("/start"));
  }

  async stop(): Promise<void> {
    await this.http.post(this.path("/stop"));
  }

  async remove(): Promise<void> {
    await this.http.delete(this.path(""));
  }

  async upload(src: Uint8Array, remotePath: string): Promise<void> {
    const qs = `?path=${encodeURIComponent(remotePath)}`;
    await this.http.requestRaw(
      "POST",
      this.path(`/files/upload${qs}`),
      src,
      "application/octet-stream",
    );
  }

  async download(remotePath: string): Promise<ArrayBuffer> {
    const qs = `?path=${encodeURIComponent(remotePath)}`;
    const res = await this.http.requestRaw("GET", this.path(`/files/download${qs}`));
    return res.arrayBuffer();
  }

  async listFiles(dirPath?: string): Promise<FileEntry[]> {
    const qs = dirPath ? `?path=${encodeURIComponent(dirPath)}` : "";
    return this.http.get(this.path(`/files${qs}`));
  }

  async listPorts(): Promise<EnvironmentPort[]> {
    return this.http.get(this.path("/ports"));
  }

  async createPort(req: CreatePortRequest): Promise<EnvironmentPort> {
    return this.http.post(this.path("/ports"), req);
  }

  async deletePort(portId: string): Promise<void> {
    await this.http.delete(this.path(`/ports/${portId}`));
  }
}

File: go/environment.go
package codewire

import (
	"context"
	"io"
	"net/url"

	"codewire.sh/sdk-go/internal/api"
)

// Environment wraps the generated api.Environment data with convenience methods.
type Environment struct {
	api.Environment
	client *Client
	orgID  string
}

func (e *Environment) Exec(ctx context.Context, req api.ExecBody) (*api.ExecResult, error) {
	var result api.ExecResult
	err := e.client.do(ctx, "POST", orgPath(e.orgID, "/environments/%s/exec", e.Id), req, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

func (e *Environment) Start(ctx context.Context) error {
	return e.client.do(ctx, "POST", orgPath(e.orgID, "/environments/%s/start", e.Id), nil, nil)
}

func (e *Environment) Stop(ctx context.Context) error {
	return e.client.do(ctx, "POST", orgPath(e.orgID, "/environments/%s/stop", e.Id), nil, nil)
}

func (e *Environment) Remove(ctx context.Context) error {
	return e.client.do(ctx, "DELETE", orgPath(e.orgID, "/environments/%s", e.Id), nil, nil)
}

func (e *Environment) Upload(ctx context.Context, src io.Reader, remotePath string) error {
	path := orgPath(e.orgID, "/environments/%s/files/upload?path=%s", e.Id, url.QueryEscape(remotePath))
	resp, err := e.client.doRaw(ctx, "POST", path, src, "application/octet-stream")
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}

func (e *Environment) Download(ctx context.Context, remotePath string) (io.ReadCloser, error) {
	path := orgPath(e.orgID, "/environments/%s/files/download?path=%s", e.Id, url.QueryEscape(remotePath))
	resp, err := e.client.doRaw(ctx, "GET", path, nil, "")
	if err != nil {
		return nil, err
	}
	return resp.Body, nil
}

func (e *Environment) ListFiles(ctx context.Context, dirPath *string) ([]api.FileEntry, error) {
	path := orgPath(e.orgID, "/environments/%s/files", e.Id)
	if dirPath != nil {
		path += "?path=" + url.QueryEscape(*dirPath)
	}
	var files []api.FileEntry
	if err := e.client.do(ctx, "GET", path, nil, &files); err != nil {
		return nil, err
	}
	return files, nil
}

func (e *Environment) ListPorts(ctx context.Context) ([]api.EnvironmentPort, error) {
	var ports []api.EnvironmentPort
	err := e.client.do(ctx, "GET", orgPath(e.orgID, "/environments/%s/ports", e.Id), nil, &ports)
	if err != nil {
		return nil, err
	}
	return ports, nil
}

func (e *Environment) CreatePort(ctx context.Context, req api.CreatePortBody) (*api.EnvironmentPort, error) {
	var port api.EnvironmentPort
	err := e.client.do(ctx, "POST", orgPath(e.orgID, "/environments/%s/ports", e.Id), req, &port)
	if err != nil {
		return nil, err
	}
	return &port, nil
}

func (e *Environment) DeletePort(ctx context.Context, portID string) error {
	return e.client.do(ctx, "DELETE", orgPath(e.orgID, "/environments/%s/ports/%s", e.Id, portID), nil, nil)
}
